From: Markus Metzger <markus.t.metzger@intel.com>
Subject: perf, nmi: Fix unknown NMI warning
Date: Fri Feb 14 16:44:08 2014 -0800
Git-commit: a3ef2229c94ff70998724cb64b9cb4c77db9e950
Patch-mainline: v3.14-rc4 
References: bsc#968512
Signed-off-by: Tony Jones <tonyj@suse.de>

    perf, nmi: Fix unknown NMI warning
    
    When using BTS on Core i7-4*, I get the below kernel warning.
    
    $ perf record -c 1 -e branches:u ls
    Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
     kernel:[  438.317893] Uhhuh. NMI received for unknown reason 31 on CPU 2.
    
    Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
     kernel:[  438.317920] Do you have a strange power saving mode enabled?
    
    Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
     kernel:[  438.317945] Dazed and confused, but trying to continue
    
    Make intel_pmu_handle_irq() take the full exit path when returning early.
    
    Cc: eranian@google.com
    Cc: peterz@infradead.org
    Cc: mingo@kernel.org
    Signed-off-by: Markus Metzger <markus.t.metzger@intel.com>
    Signed-off-by: Andi Kleen <ak@linux.intel.com>
    Signed-off-by: Peter Zijlstra <peterz@infradead.org>
    Link: http://lkml.kernel.org/r/1392425048-5309-1-git-send-email-andi@firstfloor.org
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/arch/x86/kernel/cpu/perf_event_intel.c b/arch/x86/kernel/cpu/perf_event_intel.c
index 0fa4f24..698ae77 100644
--- a/arch/x86/kernel/cpu/perf_event_intel.c
+++ b/arch/x86/kernel/cpu/perf_event_intel.c
@@ -1361,10 +1361,8 @@ static int intel_pmu_handle_irq(struct pt_regs *regs)
 	intel_pmu_disable_all();
 	handled = intel_pmu_drain_bts_buffer();
 	status = intel_pmu_get_status();
-	if (!status) {
-		intel_pmu_enable_all(0);
-		return handled;
-	}
+	if (!status)
+		goto done;
 
 	loops = 0;
 again:
